toolchain/parse/state.def

```diff
@@ -225,21 +225,26 @@ CARBON_PARSE_STATE(CodeBlock)
 //   (state done)
 CARBON_PARSE_STATE(CodeBlockFinish)
 
-// Handles a declaration name and parameters, such as `Foo[...](...)`.
+// Handles a declaration name and parameters, such as `Foo[...](...).Bar(...)`.
+// The variants impose different restrictions on the parameters:
+// - NoParams indicates that the declaration cannot contain parameters anywhere.
+// - QualifierParams indicates that the name qualifiers can have parameters,
+//   but the final name cannot (for example, `Foo[...](...).Bar` is allowed).
+// - AllParams indicates that the name can have parameters anywhere.
```

Comment:
I think Jon was saying similar: It does seem a bit surprising to be able to distinguish QualifierParams vs AllParams without knowing the type of Bar (i.e. in parse instead of check). It may be a callable type, a generic type, or not. And that may change depending on how Foo is resolved as well? Maybe there's some specific cases where Bar is fixed and this comment could explain that?